prov/efa/src/efa_rma.c

```diff
@@ -0,0 +1,417 @@
+	err = efa_rma_check_cap(base_ep);
+	if (err)
+		return err;
+
+	len = ofi_total_iov_len(iov, iov_count);
+	EFA_SETUP_RMA_IOV(rma_iov, addr, len, key);
+	EFA_SETUP_MSG_RMA(msg, iov, desc, iov_count, src_addr, &rma_iov, 1,
+			  context, 0);
+
+	return efa_rma_readmsg(ep_fid, &msg, 0);
+}
+
+static
+ssize_t efa_rma_read(struct fid_ep *ep_fid, void *buf, size_t len, void *desc,
+		     fi_addr_t src_addr, uint64_t addr, uint64_t key,
+		     void *context)
+{
+	struct iovec iov;
+	struct efa_base_ep *base_ep;
+	int err;
+
+	base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	assert(len <= base_ep->max_rma_size);
+	err = efa_rma_check_cap(base_ep);
+	if (err)
+		return err;
+
+	EFA_SETUP_IOV(iov, buf, len);
+	return efa_rma_readv(ep_fid, &iov, &desc, 1, src_addr, addr, key, context);
+}
+
+/**
+ * @brief Post a WRITE request
+ *
+ * Input:
+ *     base_ep: endpoint
+ *     msg: read operation information
+ *     flags: flags passed
+ * @return On success return 0, otherwise return a negative libfabric error code.
+ */
+static inline ssize_t efa_rma_post_write(struct efa_base_ep *base_ep,
+					 const struct fi_msg_rma *msg,
+					 uint64_t flags)
+{
+	struct efa_qp *qp;
+	struct efa_conn *conn;
+#ifndef _WIN32
+	struct ibv_sge sge_list[msg->iov_count];
+	struct ibv_data_buf inline_data_list[msg->iov_count];
+#else
+	/* MSVC compiler does not support array declarations with runtime size, so hardcode
+	 * the expected iov_limit/max_sq_sge from the lower-level efa provider.
+	 */
+	struct ibv_sge sge_list[EFA_DEV_ATTR_MAX_WR_SGE];
+	struct ibv_data_buf inline_data_list[EFA_DEV_ATTR_MAX_WR_SGE];
+#endif
+	bool self_comm;
+	size_t len;
+	int i, err = 0;
+
+	qp = base_ep->qp;
+	if (!base_ep->is_wr_started) {
+		ibv_wr_start(qp->ibv_qp_ex);
+		base_ep->is_wr_started = true;
+	}
+	qp->ibv_qp_ex->wr_id = (uintptr_t)msg->context;
+
+	if (flags & FI_REMOTE_CQ_DATA) {
+		ibv_wr_rdma_write_imm(qp->ibv_qp_ex, msg->rma_iov[0].key,
+				      msg->rma_iov[0].addr, msg->data);
+	} else {
+		ibv_wr_rdma_write(qp->ibv_qp_ex, msg->rma_iov[0].key, msg->rma_iov[0].addr);
+	}
+
+	len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+	if (len <= base_ep->domain->device->efa_attr.inline_buf_size &&
+	    len <= base_ep->inject_rma_size &&
+	    (!msg->desc || !efa_mr_is_hmem(msg->desc[0]))) {
+		for (i = 0; i < msg->iov_count; i++) {
+			inline_data_list[i].addr = msg->msg_iov[i].iov_base;
+			inline_data_list[i].length = msg->msg_iov[i].iov_len;
+		}
+		ibv_wr_set_inline_data_list(qp->ibv_qp_ex, msg->iov_count, inline_data_list);
+	} else {
+		for (i = 0; i < msg->iov_count; ++i) {
+			sge_list[i].addr = (uint64_t)msg->msg_iov[i].iov_base;
+			sge_list[i].length = msg->msg_iov[i].iov_len;
+			assert(msg->desc && msg->desc[i]);
+			sge_list[i].lkey = ((struct efa_mr *)msg->desc[i])->ibv_mr->lkey;
+		}
+		ibv_wr_set_sge_list(qp->ibv_qp_ex, msg->iov_count, sge_list);
+	}
+
+	if (msg->addr == FI_ADDR_NOTAVAIL) {
+		self_comm = true;
```

Comment:
I believe we used to allow a `FI_ADDR_NOTAVAIL` to support local read. But per libfabric standard we shouldn't allow a NOTAVAIL addr for a user call. So in that regard I think we should drop the L244-246.